dot/state/epoch.go

```diff
@@ -365,3 +506,125 @@ func (s *EpochState) SkipVerify(header *types.Header) (bool, error) {
+	defer s.nextConfigDataLock.Unlock()
+
+	_, has := s.nextConfigData[epoch]
+	if !has {
+		s.nextConfigData[epoch] = make(map[common.Hash]types.NextConfigData)
+	}
+	s.nextConfigData[epoch][hash] = nextConfigData
+}
+
+// FinalizeBABENextEpochData stores the right types.NextEpochData by
+// getting the set of hashes from the received epoch and for each hash
+// check if the header is in the database then it's been finalized and
+// thus we can also set the corresponding EpochData in the database
+func (s *EpochState) FinalizeBABENextEpochData(epoch uint64) error {
+	s.nextEpochDataLock.Lock()
+	defer s.nextEpochDataLock.Unlock()
+
+	finalizedNextEpochData, err := lookupForNextEpochPersistedHash(s.nextEpochData, s, epoch)
+	if err != nil {
+		return fmt.Errorf("cannot find next epoch data: %w", err)
+	}
+
+	ed, err := finalizedNextEpochData.ToEpochData()
+	if err != nil {
+		return fmt.Errorf("cannot transform epoch data: %w", err)
+	}
+
+	err = s.SetEpochData(epoch, ed)
+	if err != nil {
+		return fmt.Errorf("cannot set epoch data: %w", err)
+	}
+
+	// remove previous epochs from the memory
+	epochsToDelete := make([]uint64, 0)
+	for e := range s.nextEpochData {
+		if e <= epoch {
+			epochsToDelete = append(epochsToDelete, e)
+		}
+	}
+
+	for _, e := range epochsToDelete {
+		delete(s.nextEpochData, e)
+	}
+
+	return nil
+}
+
+// FinalizeBABENextConfigData stores the right types.NextConfigData by
+// getting the set of hashes from the received epoch and for each hash
+// check if the header is in the database then it's been finalized and
+// thus we can also set the corresponding NextConfigData in the database
+func (s *EpochState) FinalizeBABENextConfigData(epoch uint64) error {
+	s.nextConfigDataLock.Lock()
+	defer s.nextConfigDataLock.Unlock()
+
+	finalizedNextConfigData, err := lookupForNextEpochPersistedHash(s.nextConfigData, s, epoch)
+	if err != nil {
+		return fmt.Errorf("cannot find next config data: %w", err)
+	}
+
+	cd := finalizedNextConfigData.ToConfigData()
+	err = s.SetConfigData(epoch, cd)
+	if err != nil {
+		return fmt.Errorf("cannot set config data: %w", err)
+	}
+
+	// remove previous epochs from the memory
+	epochsToDelete := make([]uint64, 0)
+	for e := range s.nextConfigData {
+		if e <= epoch {
+			epochsToDelete = append(epochsToDelete, e)
+		}
+	}
+
+	for _, e := range epochsToDelete {
+		delete(s.nextConfigData, e)
+	}
+
+	return nil
+}
+
+// lookupForNextEpochPersistedHash given a specif epoch (the key) will go through the hashes looking
+// for a database persisted hash (belonging to the finalized chain)
+// which contains the right configuration to be persisted and safely used
+func lookupForNextEpochPersistedHash[T types.NextConfigData | types.NextEpochData](
+	nextEpochMap map[uint64]map[common.Hash]T, es *EpochState, epoch uint64) (next *T, err error) {
+	hashes, has := nextEpochMap[epoch]
+	if !has {
+		return nil, ErrEpochNotInMemory
+	}
+
+	for hash, inMemoryNextEpochConfig := range hashes {
```

Comment:
nit change `inMemoryNextEpochConfig` to something more generic... since it uses generics